src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.SyntaxAndDeclarationManager.cs

```diff
@@ -0,0 +1,587 @@
+                this.LoadedSyntaxTreeMap = loadedSyntaxTreeMap;
+                this.RootNamespaces = rootNamespaces;
+                this.DeclarationTable = declarationTable;
+            }
+
+            /// <summary>
+            /// Appends all trees (including any trees from #load'ed files).
+            /// </summary>
+            private static void AppendAllSyntaxTrees(
+                ArrayBuilder<SyntaxTree> treesBuilder,
+                SyntaxTree tree,
+                CompilationOptions options,
+                CommonMessageProvider messageProvider,
+                bool isSubmission,
+                IDictionary<SyntaxTree, int> ordinalMapBuilder,
+                IDictionary<SyntaxTree, ImmutableArray<LoadDirective>> loadDirectiveMapBuilder,
+                IDictionary<string, SyntaxTree> loadedSyntaxTreeMapBuilder,
+                IDictionary<SyntaxTree, Lazy<RootSingleNamespaceDeclaration>> declMapBuilder,
+                ref DeclarationTable declTable,
+                ref bool reuseReferenceManager)
+            {
+                var sourceCodeKind = tree.Options.Kind;
+                if (sourceCodeKind == SourceCodeKind.Interactive || sourceCodeKind == SourceCodeKind.Script)
+                {
+                    AppendAllLoadedSyntaxTrees(treesBuilder, tree, options, messageProvider, isSubmission, ordinalMapBuilder, loadDirectiveMapBuilder, loadedSyntaxTreeMapBuilder, declMapBuilder, ref declTable, ref reuseReferenceManager);
+                }
+
+                AddSyntaxTreeToDeclarationMapAndTable(tree, options, isSubmission, declMapBuilder, ref declTable, ref reuseReferenceManager);
+
+                treesBuilder.Add(tree);
+
+                ordinalMapBuilder.Add(tree, ordinalMapBuilder.Count);
+            }
+
+            private static void AppendAllLoadedSyntaxTrees(
+                ArrayBuilder<SyntaxTree> treesBuilder,
+                SyntaxTree tree,
+                CompilationOptions options,
+                CommonMessageProvider messageProvider,
+                bool isSubmission,
+                IDictionary<SyntaxTree, int> ordinalMapBuilder,
+                IDictionary<SyntaxTree, ImmutableArray<LoadDirective>> loadDirectiveMapBuilder,
+                IDictionary<string, SyntaxTree> loadedSyntaxTreeMapBuilder,
+                IDictionary<SyntaxTree, Lazy<RootSingleNamespaceDeclaration>> declMapBuilder,
+                ref DeclarationTable declTable,
+                ref bool reuseReferenceManager)
+            {
+                ArrayBuilder<LoadDirective> loadDirectives = null;
+
+                // TODO: What if SourceReferenceResolver is null?
+                var resolver = options.SourceReferenceResolver;
+                foreach (var directive in tree.GetCompilationUnitRoot().GetLoadDirectives())
+                {
+                    SyntaxTree loadedTree = null;
+                    var diagnostics = DiagnosticBag.GetInstance();
```

Comment:
Declarations can be moved below if (path == null) case below, which would also avoid the need to free the DiagnosticBag in the case.